### DISCUSSION TITLE
Make TcMenu understandable for hobbyist programmers?

### DISCUSSION THREAD

**Discussion by deruiter** [davetcc/tcMenu#418] [Q&A] [answered]
I would love to use the menu system for my open-source Fan controller, I made my own menu but changes are laborious.

But I don't know how to use the variable data generated from the menu in the sketch. 
The examples on thecoderscorner are not clear and not consistent with the screen shots so I got stuck!

Any help would be very much appreciated...

example sketch I made:
// Global Menu Item exports
extern TextMenuItem menussid;
extern EnumMenuItem menuport;
extern AnalogMenuItem menuvoltage;

![menu1](https://github.com/davetcc/tcMenu/assets/8522600/0c652802-c66e-4bb2-89ba-3d1ace09ee17)

**Comment by deruiter** [comment]
Ow, after watching the old video's I saw the solution.  

for other beginners:

`// -------------------------------------------------------------------------------
void CALLBACK_FUNCTION voltageChange(int id) {
    voltage = menuvoltage.getCurrentValue();
    Serial.println(voltage);
}


void CALLBACK_FUNCTION portChange(int id) {
    Serial.println(menuport.getCurrentValue());
}


void CALLBACK_FUNCTION ssidChange(int id) {
    Serial.println(menussid.getTextValue());
}
`

**Comment by deruiter** [comment]
For me, after another wasted day, I will quit looking into this amazing menu system until it (hopefully) will be made more clear for the average user.  A few real world examples would be so great.   I find it mind boggling that 10 years after the Arduino was made available, there is still no menu system that is usable for the hobbyist.

**Reply by davetcc** [reply]
I'm trying to understand where things have gone wrong here.

>  I find it mind boggling that 10 years after the Arduino was made available, there is still no menu system that is usable for the hobbyist.

There are many hobbyists and commercial users using tcMenu, most build reasonably complete menu solutions. They normally do have a background in C++ and are reasonably used to Arduino or other embedded development though. How used to Arduino/mbed development are you?

> A few real world examples would be so great

The TcMenu library examples folder has many real-world examples, they update fields and even cover a wide range of boards out of the box, if your hardware is reasonably commodity, you could probably get one of the examples going on a board quite quickly. They cover common screen combinations and common input plugins. From the simple Uno dfRobot example right through to mbed RTOS based examples using a screen buffer. These are packaged with the library, and in Arduino IDE and would be in the examples list under tcMenu: https://github.com/davetcc/tcMenuLib/tree/master/examples

Did you look into these examples and still decide they were not clear enough? For example this one covers the cases you are asking about and has little else within it: https://github.com/davetcc/tcMenuLib/blob/master/examples/avr/analogDfRobot/analogDfRobot.ino.

**Reply by davetcc** [reply]
I'm not trying to say the videos are perfect, they are not, but at 18:00 minutes into TcMenu 4.0 pt1 it also goes through how to acquire menu values and also to set them. It even has a chapter link in the youtube video "Looking at generated code".

**Reply by davetcc** [reply]
I mean in the simplest example, the dfRobot one, it has the following:

    // now we read the value of A0 every 200millis and set it onto a menu item
    taskManager.scheduleFixedRate(200, [] {
        menuValueA0.setCurrentValue(int(analog.getCurrentFloat(A1) * 100.0));
    });

and

    // This registers a special callback, unlike the usual one, this callback is
    // only called when editing is completely finished, you register it once and
    // it triggers for every menu item.    
    menuMgr.setItemCommittedHook([](int id) {
        menuCommits.setCurrentValue(menuCommits.getCurrentValue() + 1);
    });

Is the issue that because of the IDE you're using you didn't see the examples, or the examples in themselves are unclear? One of the issues is that PlatformIO users will not see the examples, unlike in Arduino IDE where the library examples are added directly to the examples menu.

If this is the case, could you suggest the best place to link the examples from to avoid such a situation in future?

**Comment by deruiter** [comment]
I want to thank you for your patience and the lack of my programming skills is now showing.  I have no training whatsoever, just begun to build for example the DCF clock from idea to working item but that was possible with basic c++, learning c++ on the go.

I will look into the examples! I did look into them but (stupidly) only the ESP32 ones.  Will look into the other to find an example how to use the menu ONLY when needed.  I think that is NOT the way it should work...  Thanks for pointing me to the examples.

![dcf](https://github.com/davetcc/tcMenu/assets/8522600/e7279b59-3a01-4242-b07b-c63f358fb8d0)

**Reply by davetcc** [reply, marked as answer]
No problem. I’m trying to work out what’s gone wrong here to make sure it is easier to find the right info in future. 

Methods takeOverDisplay/giveBackDisplay are shown in many of the examples. They should let you keep the menu out of the way when not in use easily enough.
